pennylane/math/ode.py

```diff
@@ -0,0 +1,243 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# TODO: This code is heavily inspired by https://github.com/google/jax/blob/main/jax/experimental/ode.py
+# need to make sure copyright is not infringed
+"""Fixed step size ODE solver"""
+
+import warnings
+from functools import partial
+
+try:
+    import jax
+    import jax.numpy as jnp
+    from jax import linear_util as lu
+    from jax.flatten_util import ravel_pytree
+    from jax.experimental import host_callback
+except ImportError as e:
+    raise ImportError(
+        "Module jax is required for ``qml.math.odeint`` functionality. "
+        "You can install jax via: pip install jax"
+    ) from e
+
+
+@partial(jax.jit, static_argnums=0)
+def odeint(func, y0, ts, *args, atol=1e-8, rtol=1e-8):
+    r"""Fix step size ODE solver with jit-compilation
+
+    Solves the initial value problem (IVP) of an ordinary differential equation (ODE)
+
+    .. math:: \frac{dy}{dt} = f(y, t), y(t_0) = y_0
+
+    using an `explicit Runge-Kutta method <https://en.wikipedia.org/wiki/Runge%E2%80%93Kutta_methods>`_.
+    In particular, it is using the 4th order term of the `Dormand Prince <https://en.wikipedia.org/wiki/Dormand%E2%80%93Prince_method>`_ Butcher table.
+
+    Args:
+        func (callable): ``f(y, t, *args)`` defining the ODE. Needs to return a ``ndarray`` of the same shape as ``y``.
+        y0 (ndarray): initial value ``y(t0) = y0``
```

Comment:
We could add a statement regarding data types, as they are handled explicitly in JAX. For example, the requirement that `func(y0, t)` does not return a changed data type object, and this may require users to cast `y0` to the right `dtype` in advance, like in the Schroedinger equation examples in the tests.

Reply:
Good point, I am a fan of being quite specific about these things in the docs and expanded here accordingly